File: expense-tracker/routes/categories.py
from flask import Blueprint, render_template, request, redirect, url_for, session, flash, jsonify
from database.db import get_db

categories_bp = Blueprint('categories', __name__)


@categories_bp.route('/categories', methods=['GET'], endpoint='list_categories')
def list_categories():
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    user_id = session['user_id']
    cats = db.execute(
        "SELECT id, name, icon, color FROM custom_categories WHERE user_id = ? ORDER BY name",
        (user_id,)
    ).fetchall()
    # Count usage per category
    for cat in cats:
        usage = db.execute(
            "SELECT COUNT(*) as cnt FROM expenses WHERE user_id = ? AND category = ?",
            (user_id, cat['name'])
        ).fetchone()['cnt'] or 0
        cat['usage_count'] = usage
    return render_template('categories.html', categories=cats)


@categories_bp.route('/categories/add', methods=['POST'], endpoint='add_category')
def add_category():
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    name = (request.form.get('name') or '').strip()
    icon = (request.form.get('icon') or 'category').strip()
    color = (request.form.get('color') or '#6366f1').strip()
    if not name or len(name) < 2:
        flash('Category name must be at least 2 characters.', 'danger')
        return redirect(url_for('categories.list_categories'))
    db = get_db()
    try:
        db.execute(
            "INSERT INTO custom_categories (user_id, name, icon, color) VALUES (?, ?, ?, ?)",
            (session['user_id'], name, icon, color)
        )
        db.commit()
        flash(f'Category "{name}" created!', 'success')
    except db.IntegrityError:
        flash(f'Category "{name}" already exists.', 'danger')
    return redirect(url_for('categories.list_categories'))


@categories_bp.route('/categories/<int:id>/edit', methods=['POST'], endpoint='edit_category')
def edit_category(id):
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    name = (request.form.get('name') or '').strip()
    icon = (request.form.get('icon') or 'category').strip()
    color = (request.form.get('color') or '#6366f1').strip()
    if not name or len(name) < 2:
        flash('Category name must be at least 2 characters.', 'danger')
        return redirect(url_for('categories.list_categories'))
    db = get_db()
    old = db.execute(
        "SELECT * FROM custom_categories WHERE id = ? AND user_id = ?",
        (id, session['user_id'])
    ).fetchone()
    if not old:
        flash('Category not found.', 'danger')
        return redirect(url_for('categories.list_categories'))
    # Update existing expenses with old name → new name
    db.execute(
        "UPDATE expenses SET category = ? WHERE user_id = ? AND category = ?",
        (name, session['user_id'], old['name'])
    )
    # Update the category itself
    db.execute(
        "UPDATE custom_categories SET name = ?, icon = ?, color = ? WHERE id = ? AND user_id = ?",
        (name, icon, color, id, session['user_id'])
    )
    db.commit()
    flash(f'Category updated to "{name}"!', 'success')
    return redirect(url_for('categories.list_categories'))


@categories_bp.route('/categories/<int:id>/delete', methods=['POST'], endpoint='delete_category')
def delete_category(id):
    if 'user_id' not in session:
        return redirect(url_for('auth.login'))
    db = get_db()
    cat = db.execute(
        "SELECT * FROM custom_categories WHERE id = ? AND user_id = ?",
        (id, session['user_id'])
    ).fetchone()
    if not cat:
        flash('Category not found.', 'danger')
        return redirect(url_for('categories.list_categories'))
    # Reassign expenses from this category to "Other"
    db.execute(
        "UPDATE expenses SET category = 'Other' WHERE user_id = ? AND category = ?",
        (session['user_id'], cat['name'])
    )
    db.execute("DELETE FROM custom_categories WHERE id = ?", (id,))
    db.commit()
    flash(f'Category "{cat["name"]}" deleted. Expenses reassigned to Other.', 'info')
    return redirect(url_for('categories.list_categories'))


@categories_bp.route('/api/categories', endpoint='api_categories')
def api_categories():
    """Return list of custom categories as JSON. Used by expense forms."""
    if 'user_id' not in session:
        return jsonify([])
    db = get_db()
    cats = db.execute(
        "SELECT id, name FROM custom_categories WHERE user_id = ? ORDER BY name",
        (session['user_id'],)
    ).fetchall()
    return jsonify([dict(c) for c in cats])

File: backend/routes/categories.js
const express = require('express');
const router = express.Router();
const mongoose = require('mongoose');
const { CustomCategory, Expense } = require('../models');
const { is_valid_hex_color } = require('../src/helpers');

router.get('/categories', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;

    try {
        const cats = await CustomCategory.find({ user_id }).sort({ name: 1 }).lean();

        const usageCounts = await Expense.aggregate([
            { $match: { user_id: new mongoose.Types.ObjectId(user_id) } },
            { $group: { _id: '$category', count: { $sum: 1 } } }
        ]);
        const usageMap = {};
        for (const uc of usageCounts) {
            usageMap[uc._id] = uc.count;
        }

        const cats_data = cats.map(cat => ({
            ...cat,
            id: cat._id.toString(),
            usage_count: usageMap[cat.name] || 0
        }));

        res.render('categories.html', { categories: cats_data });
    } catch (err) {
        next(err);
    }
});

router.post('/categories/add', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const name = (req.body.name || '').trim();
    const icon = (req.body.icon || 'category').trim();
    const color = (req.body.color || '#6366f1').trim();

    if (!name || name.length < 2) {
        req.flash('danger', 'Category name must be at least 2 characters.');
        return res.redirect('/categories');
    }

    if (!is_valid_hex_color(color)) {
        req.flash('danger', 'Invalid category color.');
        return res.redirect('/categories');
    }

    try {
        const newCat = new CustomCategory({
            user_id,
            name,
            icon,
            color
        });
        await newCat.save();
        req.flash('success', `Category "${name}" created!`);
    } catch (err) {
        if (err.code === 11000) {
            req.flash('danger', `Category "${name}" already exists.`);
        } else {
            return next(err);
        }
    }
    res.redirect('/categories');
});

router.post('/categories/:id/edit', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const cat_id = req.params.id;

    if (!mongoose.isValidObjectId(cat_id)) {
        req.flash('danger', 'Category not found.');
        return res.redirect('/categories');
    }

    const name = (req.body.name || '').trim();
    const icon = (req.body.icon || 'category').trim();
    const color = (req.body.color || '#6366f1').trim();

    if (!name || name.length < 2) {
        req.flash('danger', 'Category name must be at least 2 characters.');
        return res.redirect('/categories');
    }

    if (!is_valid_hex_color(color)) {
        req.flash('danger', 'Invalid category color.');
        return res.redirect('/categories');
    }

    try {
        const oldCat = await CustomCategory.findOne({ _id: cat_id, user_id });
        if (!oldCat) {
            req.flash('danger', 'Category not found.');
            return res.redirect('/categories');
        }

        // Update all expenses with old name to the new name
        await Expense.updateMany(
            { user_id: new mongoose.Types.ObjectId(user_id), category: oldCat.name },
            { category: name }
        );

        oldCat.name = name;
        oldCat.icon = icon;
        oldCat.color = color;
        await oldCat.save();

        req.flash('success', `Category updated to "${name}"!`);
    } catch (err) {
        if (err.code === 11000) {
            req.flash('danger', `Category "${name}" already exists.`);
        } else {
            return next(err);
        }
    }
    res.redirect('/categories');
});

router.post('/categories/:id/delete', async (req, res, next) => {
    if (!req.session.user_id) {
        return res.redirect('/login');
    }
    const user_id = req.session.user_id;
    const cat_id = req.params.id;

    if (!mongoose.isValidObjectId(cat_id)) {
        req.flash('danger', 'Category not found.');
        return res.redirect('/categories');
    }

    try {
        const cat = await CustomCategory.findOne({ _id: cat_id, user_id });
        if (!cat) {
            req.flash('danger', 'Category not found.');
            return res.redirect('/categories');
        }

        // Reassign expenses from this category to "Other"
        await Expense.updateMany(
            { user_id: new mongoose.Types.ObjectId(user_id), category: cat.name },
            { category: 'Other' }
        );

        await CustomCategory.deleteOne({ _id: cat_id, user_id: new mongoose.Types.ObjectId(user_id) });
        req.flash('info', `Category "${cat.name}" deleted. Expenses reassigned to Other.`);
    } catch (err) {
        return next(err);
    }
    res.redirect('/categories');
});

router.get('/api/categories', async (req, res) => {
    if (!req.session.user_id) {
        return res.json([]);
    }
    try {
        const cats = await CustomCategory.find({ user_id: req.session.user_id }).sort({ name: 1 }).lean();
        res.json(cats.map(c => ({
            id: c._id.toString(),
            name: c.name
        })));
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

module.exports = router;
